test/unit/language-extensions/range.spec.ts

```diff
@@ -0,0 +1,68 @@
+import * as path from "path";
+import * as util from "../../util";
+import * as tstl from "../../../src";
+import { invalidRangeControlVariable, invalidRangeUse } from "../../../src/transformation/utils/diagnostics";
+
+const rangeProjectOptions: tstl.CompilerOptions = {
+    types: [path.resolve(__dirname, "../../../language-extensions")],
+};
+
+test("$range basic use", () => {
+    util.testFunction`
+		const result: number[] = [];
+		for (const i of $range(1, 5)) {
+			result.push(i);
+		}
+		return result;
+	`
+        .setOptions(rangeProjectOptions)
+        .expectToEqual([1, 2, 3, 4, 5]);
+});
+
+test("$range basic use with step", () => {
+    util.testFunction`
+		const result: number[] = [];
+		for (const i of $range(1, 10, 2)) {
+			result.push(i);
+		}
+		return result;
+	`
+        .setOptions(rangeProjectOptions)
+        .expectToEqual([1, 3, 5, 7, 9]);
+});
+
+test("$range basic use reverse", () => {
+    util.testFunction`
+		const result: number[] = [];
+		for (const i of $range(5, 1, -1)) {
+			result.push(i);
+		}
+		return result;
+	`
+        .setOptions(rangeProjectOptions)
+        .expectToEqual([5, 4, 3, 2, 1]);
+});
+
+test("$range invalid control variable", () => {
+    util.testFunction`
+		let i: number;
+		for (i of $range(1, 5)) {}
+	`
+        .setOptions(rangeProjectOptions)
+        .expectDiagnosticsToMatchSnapshot([invalidRangeControlVariable.code]);
+});
+
+test.each([
+    ["for (const i in $range(1, 10, 2)) {}", [invalidRangeUse.code]],
+    ["const x = $range(1, 10);", [invalidRangeUse.code]],
+    ["const range = $range;", [invalidRangeUse.code, invalidRangeUse.code]],
```

Comment:
Same code mentioned twice here, could that be reduced to 1? Then is there a need to specify them as part of the each() at all or could you just put it in the test body?

Reply:
This is an odd side effect of how the $range function was being detected. I found a way to do that better and avoid the redundant diagnostic.